pr_agent/algo/ai_handler.py

```diff
@@ -92,17 +105,53 @@ async def chat_completion(self, model: str, system: str, user: str, temperature:
                     f"Generating completion with {model}"
                     f"{(' from deployment ' + deployment_id) if deployment_id else ''}"
                 )
-            response = await acompletion(
-                model=model,
-                deployment_id=deployment_id,
-                messages=[
-                    {"role": "system", "content": system},
-                    {"role": "user", "content": user}
-                ],
-                temperature=temperature,
-                azure=self.azure,
-                force_timeout=get_settings().config.ai_timeout
-            )
+            
+            bito_response = None
+            response = None
+            if model.startswith("bito"):
+                global trying
+                trying += 1
+                logging.info(f"trying {trying}/{OPENAI_RETRIES} to get AI prediction...")
+                model_type = model.replace("bito_", "").upper()
+                tmp = tempfile.NamedTemporaryFile(mode='w', delete=False) 
+                context_file = tmp.name
+
+                out = tempfile.NamedTemporaryFile(mode='w+t', delete=False)
+                out_file = out.name
+
+                with tempfile.NamedTemporaryFile(mode='w', delete=False) as system_prompt:
+                    system_prompt.write(system)
+                    instruction_file = system_prompt.name
+                with tempfile.NamedTemporaryFile(mode='w', delete=False) as user_prompt:
+                    user_prompt.write(user)
+                    cmd_output_file = user_prompt.name
+                bito_cmd = f"bito -p '{instruction_file}' -c '{context_file} -k {self.bito_cli_key}'"
+                bito_cmd += f" -f '{cmd_output_file}'" # Pass user prompt in this file
+                bito_cmd += f" -m '{model_type}'"
+                bito_cmd += f" > {out_file}"
+                proc = await asyncio.create_subprocess_shell(
+                    bito_cmd, stdout=None, 
+                    stderr=subprocess.PIPE
+                )
+                _, stderr = await proc.communicate()
+                if stderr:
+                    logging.info(f"Error from bito cli: {stderr.decode('utf-8')}")
+                else:
+                    with open(out_file, 'r') as file:
+                        bito_response = file.read()    
+                # bito_response = stdout.decode('utf-8')
+            else:
+                response = await acompletion(
+                    model=model,
+                    deployment_id=deployment_id,
+                    messages=[
+                        {"role": "system", "content": system},
+                        {"role": "user", "content": user}
+                    ],
+                    temperature=temperature,
+                    azure=self.azure,
+                    force_timeout=get_settings().config.ai_timeout
```

Comment:
**Performance Issue**: The introduction of a conditional flow to handle different AI models based on the prefix "bito" introduces potential inefficiencies and complexities in managing subprocesses and temporary files for AI predictions. <br> **Fix**: Refactor the AI prediction handling to abstract the model invocation into a separate, optimized service layer that can manage model-specific optimizations and resource cleanup more effectively. This approach can reduce the overhead of creating and managing temporary files and subprocesses directly within the AI handler. <br> **Code Suggestion**: 
 ```
 class AiHandler:
     def __init__(self):
         self.model_service = ModelService()

     async def chat_completion(self, model: str, system: str, user: str, temperature: float):
         return await self.model_service.predict(model, system, user, temperature)

 class ModelService:
     async def predict(self, model: str, system: str, user: str, temperature: float):
         # Model prediction logic here
         pass
 ```